livox_ros_driver/livox_ros_driver/livox_ros_driver_nodelet.cpp

```diff
@@ -0,0 +1,219 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+//
+
+#include "include/livox_ros_driver_nodelet.h"
+#include <pluginlib/class_list_macros.h>
+
+#include <chrono>
+#include <vector>
+
+#include "lddc.h"
+#include "lds_hub.h"
+#include "lds_lidar.h"
+#include "lds_lvx.h"
+#include "livox_sdk.h"
+#include <ros/ros.h>
+
+namespace livox_ros
+{
+
+  const int32_t kSdkVersionMajorLimit = 2;
+
+  LivoxRosDriverNodelet::LivoxRosDriverNodelet()
+  {
+  }
+
+  LivoxRosDriverNodelet::~LivoxRosDriverNodelet()
+  {
+    if (thread_->joinable())
+    {
+      thread_->join();
+    }
+  }
+
+  void LivoxRosDriverNodelet::onInit(void)
+  {
+    n_ = getNodeHandle();
+    np_ = getPrivateNodeHandle();
+
+    ROS_INFO("Livox Ros Driver Version: %s", LIVOX_ROS_DRIVER_VERSION_STRING);
+
+    /** Ros related */
+    if (ros::console::set_logger_level(ROSCONSOLE_DEFAULT_NAME,
+                                       ros::console::levels::Debug))
+    {
+      ros::console::notifyLoggerLevelsChanged();
+    }
+    /** Check sdk version */
+    LivoxSdkVersion _sdkversion;
+    GetLivoxSdkVersion(&_sdkversion);
+    if (_sdkversion.major < kSdkVersionMajorLimit)
+    {
+      ROS_INFO("The SDK version[%d.%d.%d] is too low", _sdkversion.major,
+               _sdkversion.minor, _sdkversion.patch);
+      return;
+    }
+
+    /** Init defualt system parameter */
+    int xfer_format = kPointCloud2Msg;
+    int multi_topic = 0;
+    int data_src = kSourceRawLidar;
+    double publish_freq = 10.0; /* Hz */
+    int output_type = kOutputToRos;
+    std::string lidar_frame_id = "livox_frame";
+    std::string imu_frame_id = "livox_frame";
+    bool lidar_bag = true;
+    bool imu_bag   = false;
+
+    np_.getParam("xfer_format", xfer_format);
+    np_.getParam("multi_topic", multi_topic);
+    np_.getParam("data_src", data_src);
+    np_.getParam("publish_freq", publish_freq);
+    np_.getParam("output_data_type", output_type);
+    np_.getParam("lidar_frame_id", lidar_frame_id);
+    np_.getParam("imu_frame_id", imu_frame_id);
+    np_.getParam("enable_lidar_bag", lidar_bag);
+    np_.getParam("enable_imu_bag", imu_bag);
+
+    // Clamp publish_freq between 1.0 and 100.0 Hz
+    publish_freq = std::max(std::min(publish_freq, 100.0), 1.0);
+
+    /** Lidar data distribute control and lidar data source set */
+    lddc_= new Lddc(xfer_format, multi_topic, data_src, output_type,
+                    publish_freq, lidar_frame_id, imu_frame_id,
+                    lidar_bag, imu_bag);
+    lddc_->SetRosNode(&np_);
+
+    int ret = 0;
+    if (data_src == kSourceRawLidar)

```

Comment:
I assume that this logic was directly yanked from the ROS node?

Reply:
The code was cherry-picked from a fork of the original code so I didn't really test it since I don't have a Livox Lidar, but there may be some things to modify.

Reply:
I'll compare `livox_ros_driver.cpp` with this file.